internal/sftp/client.go

```diff
@@ -5,6 +5,22 @@ import (
 	"io"
 )
 
+type AuthError struct {
+	err error
+}
+
+func (e *AuthError) Error() string {
+	return e.err.Error()
+}
+
+func (e *AuthError) Unwrap() error {
+	return e.err
+}
+
+func NewAuthError(e error) *AuthError {
+	return &AuthError{err: e}
+}
```

Comment:
A possible alternative:

```
type AuthError struct {
	Reason string
}

func (e *AuthError) Error() string {
	return fmt.Sprintf("auth: %s", e.Reason)
}

func NewAuthError(err error) error {
	return AuthError{Reason: err.String()}
}
```

This version:
* Doesn't use Unwrap because we don't need to give access to the underlying error.
* Uses the error type in the constructor signature, see https://go.dev/doc/faq#nil_error:
   > It's a good idea for functions that return errors always to use the error type in their signature rather than a concrete type such as *MyError, to help guarantee the error is created correctly. As an example, [os.Open](https://go.dev/pkg/os/#Open) returns an error even though, if not nil, it's always of concrete type [*os.PathError](https://go.dev/pkg/os/#PathError).

I think that we'd want the function to return a concrete type (*AuthError) if we were doing some performance-focused work, e.g. https://github.com/connectrpc/connect-go/blob/d88758dc0e89170db922ecd20f16cec57662ec23/error.go#L117-L128 could be a good example where the design was driven by performance objectives, like avoiding memory allocations. This is a guess though, not entirely sure.

Reply:
@sevein does `Reason` need to be exported?

Reply:
Exporting the `Reason` field does have the nice benefit of being able to set an arbitrary string without having to create an error first, e.g. 

```
e := &sftp.AuthError{Reason: "bad password"}
```
vs.

```
e := sftp.NewAuthError(errors.New("bad password"))
```